### PR TITLE
feat: [CI-4660]: OAuth integration in CI Hosted Builds 1st time setup

### PR DESCRIPTION
**Pull request by vardanbansal-harness**
##### Summary:

In this change, we've added capability to create an SCM connector via OAuth.

https://user-images.githubusercontent.com/48940704/174798506-f2793607-c307-441f-8924-1c5b501e2218.mov

##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx [hunk 1]
```diff
@@ -105,9 +114,97 @@ const SelectGitProviderRef = (
   const [connector, setConnector] = useState<ConnectorInfoDTO>()
   const [secret, setSecret] = useState<SecretDTOV2>()
   const scrollRef = useRef<Element>()
+  const oAuthSecretIntercepted = useRef<boolean>(false)
```
Comment (d1wilko) [inline src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:117]: should this be useState? - or is the point to not rerender?

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx [hunk 2]
```diff
@@ -105,9 +114,97 @@ const SelectGitProviderRef = (
...
+      setOAuthStatus(Status.FAILURE)
+      return
+    }
+    const { accessTokenRef, refreshTokenRef, status, errorMessage } = event.data
+    //safeguard against backend server sending multiple oauth events, which could lead to multiple duplicate connectors getting created
+    if (!oAuthSecretIntercepted.current) {
+      if ((status as string).toLowerCase() === Status.SUCCESS.toLowerCase()) {
+        oAuthSecretIntercepted.current = true
+        createOAuthConnector({ tokenRef: accessTokenRef, refreshTokenRef })
+      } else if (errorMessage !== PLACEHOLDER) {
```
Comment (d1wilko) [inline src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:153]: happy with keeping this piece in?

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx [hunk 3]
```diff
@@ -757,35 +858,58 @@ const SelectGitProviderRef = (
...
+                              // ||
+                              // (gitProvider?.type === Connectors.GITHUB && oAuthStatus === Status.IN_PROGRESS)
+                            }
+                            tooltipProps={
+                              disableOAuthForGitProvider
+                                ? {
+                                    isDark: true
+                                  }
+                                : { isOpen: false }
+                            }
```
Comment (d1wilko) [inline src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:895]: is this logic right? dark OR open?

**Review by d1wilko — APPROVED** [review]
Looks good - some questions it would be good to clear uo, but still approving :)

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx (hunk 1 above)
Comment (vardanbansal-harness) [reply src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:117]: Yeah usually I try not to put variables not involved in changing view as state variables.

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx (hunk 2 above)
Comment (vardanbansal-harness) [reply src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:153]: Yeah so this is like a temporary solution. These variables `accessTokenRef`, `refreshTokenRef`, `status` and `errorMessage` come with a correct value or as a placeholder value `"placeholder"`. So in this case, I want to make sure only when `errorMessage` equals `"placeholder"`, we mark it as a failure.

File: src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx (hunk 3 above)
Comment (d1wilko) [reply src/modules/75-ci/pages/get-started-with-ci/InfraProvisioningWizard/SelectGitProvider.tsx:895]: ah fair enough - I misread the logic 🙈

**Pull request merged**